File: game/atc-core/server/runtime_sustainment.lua
local BASE = 'http://localhost:30120/api/v1/runtime-sustainment'

-- ── Runtime Sustainment ──────────────────────────────────────────────────────

AddEventHandler('atc:sustainment:initiate', function(data, cb)
  PerformHttpRequest(BASE, function(status, body)
    if cb then cb(status, body) end
  end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:start', function(id, cb)
  PerformHttpRequest(BASE .. '/' .. id .. '/start', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:maintain', function(id, cb)
  PerformHttpRequest(BASE .. '/' .. id .. '/maintain', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:complete', function(id, cb)
  PerformHttpRequest(BASE .. '/' .. id .. '/complete', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:fail', function(id, cb)
  PerformHttpRequest(BASE .. '/' .. id .. '/fail', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

-- ── Infinite Recovery ────────────────────────────────────────────────────────

AddEventHandler('atc:sustainment:recovery:initiate', function(data, cb)
  PerformHttpRequest(BASE .. '/recovery', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:recovery:begin', function(recoveryId, cb)
  PerformHttpRequest(BASE .. '/recovery/' .. recoveryId .. '/begin', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:recovery:complete', function(recoveryId, cb)
  PerformHttpRequest(BASE .. '/recovery/' .. recoveryId .. '/complete', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:recovery:fail', function(recoveryId, cb)
  PerformHttpRequest(BASE .. '/recovery/' .. recoveryId .. '/fail', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

-- ── Autonomous Maintenance ───────────────────────────────────────────────────

AddEventHandler('atc:sustainment:maintenance:schedule', function(data, cb)
  PerformHttpRequest(BASE .. '/maintenance', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:maintenance:run', function(id, cb)
  PerformHttpRequest(BASE .. '/maintenance/' .. id .. '/run', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:maintenance:complete', function(id, cb)
  PerformHttpRequest(BASE .. '/maintenance/' .. id .. '/complete', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:maintenance:skip', function(id, cb)
  PerformHttpRequest(BASE .. '/maintenance/' .. id .. '/skip', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

-- ── Distributed Sustainment Nodes ────────────────────────────────────────────

AddEventHandler('atc:sustainment:node:register', function(data, cb)
  PerformHttpRequest(BASE .. '/node', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:node:degrade', function(sustainmentNodeId, cb)
  PerformHttpRequest(BASE .. '/node/' .. sustainmentNodeId .. '/degrade', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:node:recover', function(sustainmentNodeId, cb)
  PerformHttpRequest(BASE .. '/node/' .. sustainmentNodeId .. '/recover', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:node:fail', function(sustainmentNodeId, cb)
  PerformHttpRequest(BASE .. '/node/' .. sustainmentNodeId .. '/fail', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

-- ── Runtime Longevity ────────────────────────────────────────────────────────

AddEventHandler('atc:sustainment:longevity:create', function(data, cb)
  PerformHttpRequest(BASE .. '/longevity', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:longevity:activate', function(id, cb)
  PerformHttpRequest(BASE .. '/longevity/' .. id .. '/activate', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:longevity:archive', function(id, cb)
  PerformHttpRequest(BASE .. '/longevity/' .. id .. '/archive', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sustainment:longevity:expire', function(id, cb)
  PerformHttpRequest(BASE .. '/longevity/' .. id .. '/expire', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', '{}', { ['Content-Type'] = 'application/json' })
end)

-- ── Cleanup ──────────────────────────────────────────────────────────────────

AddEventHandler('atc:sustainment:cleanup', function(thresholdMs, cb)
  PerformHttpRequest(BASE .. '/cleanup', function(status, body)
    if cb then cb(status, body) end
  end, 'POST', json.encode({ thresholdMs = thresholdMs }), { ['Content-Type'] = 'application/json' })
end)

CreateThread(function()
  while true do
    Wait(300000)
    TriggerEvent('atc:sustainment:cleanup', 300000)
  end
end)

File: apps/api/src/routes/runtime-sustainment.ts
import type { FastifyInstance } from 'fastify'
import type { AppContext } from '../context.js'
import {
  initiateRuntimeSustainmentSchema,
  initiateInfiniteRecoverySchema,
  scheduleAutonomousMaintenanceSchema,
  registerSustainmentNodeSchema,
  createRuntimeLongevitySchema,
  cleanupRuntimeSustainmentSchema,
} from '@atc/operations'

export function runtimeSustainmentRoutes(fastify: FastifyInstance, { ctx }: { ctx: AppContext }) {
  // ── Runtime Sustainment ──────────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sustainment', async (req, reply) => {
    if (!ctx.runtimeSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = initiateRuntimeSustainmentSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { sustainmentData, ...rest } = parsed.data
    const record = await ctx.runtimeSustainmentService.initiateSustainment({
      ...rest,
      ...(sustainmentData !== undefined ? { sustainmentData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/:id/start', async (req, reply) => {
    if (!ctx.runtimeSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSustainmentService.startSustainment(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/:id/maintain', async (req, reply) => {
    if (!ctx.runtimeSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSustainmentService.maintainSustainment(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/:id/complete', async (req, reply) => {
    if (!ctx.runtimeSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSustainmentService.completeSustainment(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/:id/fail', async (req, reply) => {
    if (!ctx.runtimeSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSustainmentService.failSustainment(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sustainment/:id', async (req, reply) => {
    if (!ctx.runtimeSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSustainmentService.getSustainment(id)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Infinite Recovery ────────────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sustainment/recovery', async (req, reply) => {
    if (!ctx.infiniteRecoveryCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = initiateInfiniteRecoverySchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { recoveryData, ...rest } = parsed.data
    const record = await ctx.infiniteRecoveryCoordinator.initiateRecovery({
      ...rest,
      ...(recoveryData !== undefined ? { recoveryData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/recovery/:recoveryId/begin', async (req, reply) => {
    if (!ctx.infiniteRecoveryCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const { recoveryId } = req.params as { recoveryId: string }
    const record = await ctx.infiniteRecoveryCoordinator.beginRecovering(recoveryId)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/recovery/:recoveryId/complete', async (req, reply) => {
    if (!ctx.infiniteRecoveryCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const { recoveryId } = req.params as { recoveryId: string }
    const record = await ctx.infiniteRecoveryCoordinator.completeRecovery(recoveryId)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/recovery/:recoveryId/fail', async (req, reply) => {
    if (!ctx.infiniteRecoveryCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const { recoveryId } = req.params as { recoveryId: string }
    const record = await ctx.infiniteRecoveryCoordinator.failRecovery(recoveryId)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sustainment/recovery/:recoveryId', async (req, reply) => {
    if (!ctx.infiniteRecoveryCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const { recoveryId } = req.params as { recoveryId: string }
    const record = await ctx.infiniteRecoveryCoordinator.getRecovery(recoveryId)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Autonomous Maintenance ───────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sustainment/maintenance', async (req, reply) => {
    if (!ctx.autonomousMaintenanceService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = scheduleAutonomousMaintenanceSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { maintenanceData, ...rest } = parsed.data
    const record = await ctx.autonomousMaintenanceService.scheduleMaintenance({
      ...rest,
      ...(maintenanceData !== undefined ? { maintenanceData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/maintenance/:id/run', async (req, reply) => {
    if (!ctx.autonomousMaintenanceService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousMaintenanceService.runMaintenance(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/maintenance/:id/complete', async (req, reply) => {
    if (!ctx.autonomousMaintenanceService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousMaintenanceService.completeMaintenance(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/maintenance/:id/skip', async (req, reply) => {
    if (!ctx.autonomousMaintenanceService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousMaintenanceService.skipMaintenance(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sustainment/maintenance/:id', async (req, reply) => {
    if (!ctx.autonomousMaintenanceService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousMaintenanceService.getMaintenance(id)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Distributed Sustainment Nodes ────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sustainment/node', async (req, reply) => {
    if (!ctx.distributedSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = registerSustainmentNodeSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { nodeData, ...rest } = parsed.data
    const record = await ctx.distributedSustainmentService.registerNode({
      ...rest,
      ...(nodeData !== undefined ? { nodeData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/node/:sustainmentNodeId/degrade', async (req, reply) => {
    if (!ctx.distributedSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { sustainmentNodeId } = req.params as { sustainmentNodeId: string }
    const record = await ctx.distributedSustainmentService.degradeNode(sustainmentNodeId)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/node/:sustainmentNodeId/recover', async (req, reply) => {
    if (!ctx.distributedSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { sustainmentNodeId } = req.params as { sustainmentNodeId: string }
    const record = await ctx.distributedSustainmentService.recoverNode(sustainmentNodeId)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/node/:sustainmentNodeId/fail', async (req, reply) => {
    if (!ctx.distributedSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { sustainmentNodeId } = req.params as { sustainmentNodeId: string }
    const record = await ctx.distributedSustainmentService.failNode(sustainmentNodeId)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sustainment/node/:sustainmentNodeId', async (req, reply) => {
    if (!ctx.distributedSustainmentService) return reply.code(503).send({ error: 'Service unavailable' })
    const { sustainmentNodeId } = req.params as { sustainmentNodeId: string }
    const record = await ctx.distributedSustainmentService.getNode(sustainmentNodeId)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Runtime Longevity ────────────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sustainment/longevity', async (req, reply) => {
    if (!ctx.runtimeLongevityService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = createRuntimeLongevitySchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { longevityData, ...rest } = parsed.data
    const record = await ctx.runtimeLongevityService.createCheckpoint({
      ...rest,
      ...(longevityData !== undefined ? { longevityData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/longevity/:id/activate', async (req, reply) => {
    if (!ctx.runtimeLongevityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeLongevityService.activateCheckpoint(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/longevity/:id/archive', async (req, reply) => {
    if (!ctx.runtimeLongevityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeLongevityService.archiveCheckpoint(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sustainment/longevity/:id/expire', async (req, reply) => {
    if (!ctx.runtimeLongevityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeLongevityService.expireCheckpoint(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sustainment/longevity/:id', async (req, reply) => {
    if (!ctx.runtimeLongevityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeLongevityService.getCheckpoint(id)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Cleanup ──────────────────────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sustainment/cleanup', async (req, reply) => {
    if (!ctx.sustainmentRecoveryService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = cleanupRuntimeSustainmentSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const result = await ctx.sustainmentRecoveryService.cleanupStale(parsed.data.thresholdMs)
    return reply.code(200).send(result)
  })
}
